lsp-types/src/Language/LSP/MetaModel/CodeGen.hs

```diff
@@ -0,0 +1,911 @@
+  pure $ TH.NewtypeD [] tn [] Nothing (TH.NormalC ctor [(noBang, rhs)]) [stockDeriving, newtypeDeriving]
+  --pure $ TH.TySynD tn [] rhs
+
+(.=?) :: (J.KeyValue kv, J.ToJSON v) => J.Key -> Maybe v -> [kv]
+k .=? v = case v of
+  Just v' -> [k J..= v']
+  Nothing -> mempty
+
+-- | Generate a type declaration corresponding to a top-level named struct.
+genStruct :: Structure -> CodeGenM [TH.Dec]
+genStruct s@Structure{name, documentation, since, proposed} = do
+  let structName = name
+
+  st <- asks symbolTable
+  tn <- case Map.lookup name st of
+    Just n  -> pure n
+    Nothing -> fail $ "no name for " ++ show name
+  addDocumentation tn documentation since proposed
+
+  ctor <- lift $ makeToplevelName name
+
+  props <- getStructProperties s
+  args <- for props $ \Property{name, type_, optional, documentation, since, proposed} -> do
+      pty <- convertType type_
+      let mty = case optional of
+            Just True -> TH.ConT ''Maybe `TH.AppT` pty
+            _         -> pty
+      n <- makeFieldName name
+      addDocumentation n documentation since proposed
+      pure (name, (n, noBang, mty))
+
+  let datad = TH.DataD [] tn [] Nothing [TH.RecC ctor (fmap snd args)] [stockDeriving]
+
+  toJsonD <- do
+    (unzip -> (args, pairEs)) <- for props $ \Property{name, optional} -> do
+      n <- lift $ TH.newName "arg"
+      pairE <- case optional of
+        Just True -> lift [| $(TH.litE $ TH.stringL $ T.unpack name) .=? $(TH.varE n) |]
+        _ -> lift [| [$(TH.litE $ TH.stringL $ T.unpack name) J..= $(TH.varE n)] |]
+      pure (n, pairE)
+    body <- lift [| J.object $ concat $ $(TH.listE (fmap pure pairEs)) |]
+    let toJsonF = TH.FunD 'J.toJSON [TH.Clause [TH.ConP ctor (fmap TH.VarP args)] (TH.NormalB body) []]
+    pure $ TH.InstanceD Nothing [] (TH.ConT ''J.ToJSON `TH.AppT` TH.ConT tn) [toJsonF]
+
+  fromJsonD <- do
+    vn <- lift $ TH.newName "v"
+    exprs <- for props $ \Property{name, optional} ->
+      case optional of
+        Just True -> lift [| $(TH.varE vn) J..:! $(TH.litE $ TH.stringL $ T.unpack name)|]
+        _         -> lift [| $(TH.varE vn) J..: $(TH.litE $ TH.stringL $ T.unpack name)|]
+    let lamBody = mkIterApplicativeApp (TH.ConE ctor) exprs
+    body <- lift [| J.withObject $(TH.litE $ TH.stringL $ T.unpack structName) $ \ $(TH.varP vn) -> $lamBody |]
+    let fromJsonF = TH.FunD 'J.parseJSON [TH.Clause [] (TH.NormalB body) []]
+    pure $ TH.InstanceD Nothing [] (TH.ConT ''J.FromJSON `TH.AppT` TH.ConT tn) [fromJsonF]
+
+  pure [datad, toJsonD, fromJsonD]
+
+-- | Get the list of properties of a struct, including inherited ones.
+getStructProperties :: Structure -> CodeGenM [Property]
+getStructProperties s@Structure{name, properties, extends, mixins} = do
+  st <- asks structTable
+  let
+    extends' = fromMaybe [] extends
+    mixins' = fromMaybe [] mixins
+    supertypes = extends' ++ mixins'
```

Comment:
I wonder if it's ever possible for an `extended type` or `mixin` type to have duplicate fields? Is it worth de-duping `superProps` (aside from local name de-duping)